go/vt/vtgate/planbuilder/operator_transformers.go

```diff
@@ -363,7 +363,10 @@ func transformProjection(ctx *plancontext.PlanningContext, op *operators.Project
 	if cols, colNames := op.AllOffsets(); cols != nil {
 		// if all this op is doing is passing through columns from the input, we
 		// can use the faster SimpleProjection
-		return useSimpleProjection(cols, colNames, src)
+		if len(op.Source.GetColumns(ctx)) == len(cols) && offsetInInputOrder(cols) {
+			cols = nil
+		}
+		return newSimpleProjection(cols, colNames, src)
```

Comment:
I think we can still do better here by changing a little more on simple projection like checking `len(op.Source.GetColumns(ctx)) >= len(cols)`. And Simple project truncates from the end.

can leave a TODO here if you think it make sense.

Reply:
the columns need to be of the expected size. we should not return more columns than what the user asked for

Reply:
yes, we can do again minimal work of truncate. not important for this PR